src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -131,15 +131,69 @@ internal ProjectRootElementCache(bool autoReloadFromDisk, bool loadProjectsReadO
             LoadProjectsReadOnly = loadProjectsReadOnly;
         }
 
+
+        /// <summary>
+        /// Returns true if given cache entry exists and is outdated.
+        /// </summary>
+        private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootElement)
+        {
+            if (projectRootElement != null && _autoReloadFromDisk)
+            {
+                FileInfo fileInfo = FileUtilities.GetFileInfoNoThrow(projectFile);
+
+                // If the file doesn't exist on disk, go ahead and use the cached version.
+                // It's an in-memory project that hasn't been saved yet.
+                if (fileInfo != null)
+                {
+                    if (fileInfo.LastWriteTime != projectRootElement.LastWriteTimeWhenRead)
+                    {
+                        // File was changed on disk by external means. Cached version is no longer valid.
+                        // We could throw here or ignore the problem, but it is a common and reasonable pattern to change a file
+                        // externally and load a new project over it to see the new content. So we dump it from the cache
+                        // to force a load from disk. There might then exist more than one ProjectRootElement with the same path,
+                        // but clients ought not get themselves into such a state - and unless they save them to disk,
+                        // it may not be a problem.
+                        return true;
+                    }
+                    else if (!String.IsNullOrEmpty(Environment.GetEnvironmentVariable("MSBUILDCACHECHECKFILECONTENT")))
+                    {
+                        lock (_locker)
+                        {
+                            // QA tests run too fast for the timestamp check to work. This environment variable is for their
+                            // use: it checks the file content as well as the timestamp. That's better than completely disabling
+                            // the cache as we get test coverage of the rest of the cache code.
+                            XmlDocument document = new XmlDocument();
+                            document.PreserveWhitespace = projectRootElement.XmlDocument.PreserveWhitespace;
+
+                            using (var xtr = XmlReaderExtension.Create(projectRootElement.FullPath, projectRootElement.ProjectRootElementCache.LoadProjectsReadOnly))
+                            {
+                                document.Load(xtr.Reader);
```

Comment:
This is inside the lock, right? Is that ok since it's a "test-only" codepath, or can this be avoided too?

Reply:
Since it is really used only in QA tests we haven't give it much thinking as for optimization. Shall we?

Reply:
This section is thread safe and the lock could be omitted here. This could result in multiple tests concurrently loading project XML in readonly fashion. This shall be OK and we should consider it. Worst could happen some test fails or become flaky in which situation we can put the lock back. What do you think @AR-May ?

Reply:
Yes, I think we can try to unlock this section, I suppose.

Reply:
If it's test only I think a bit of lock contention is fine to save development effort. Just wanted to make sure that was the case.